elastic/kibana
# [Serverless][Security Solution][Endpoint] Restrict endpoint exceptions on serverless via plugin sub-features

### What this PR changes

branched from elastic/kibana/pull/163759

- Introduces new AppFeatures package `@kbn/security-solution-features` with the common logic and `AppFeatureService` to apply offering specific configurations for Security Solution features independently for Serverless and ESS.  This logic is replacing the earlier `AppFeatures` in order to introduce new Kibana feature privileges for serverless PLIs so that new Kibana privileges introduced for serverless PLIs do not affect/show up as new Kibana feature privileges in ESS.
- Gates endpoint exceptions on alerts/rules based on serverless PLI configurations. On serverless `Endpoint exceptions` should be accessible/seen only on endpoint essentials/complete.

New AppFeatures logic architecture diagram:

![Security Solution Features (Current)](https://github.com/elastic/kibana/assets/17747913/f627406d-43bc-4db5-93b1-4e43eeb6d870)

**Note:** Corresponding API changes related to endpoint exceptions will be in a new PR, along with the last set of UX changes for hiding the `Endpoint exceptions` tab from the Rules details page.

### How to review

- Setup for _Servlerless_
  - Run `yarn es snapshot` on a terminal window to start ES.
  - Copy `config/serverless.security.yml` to `config/serverless.security.dev.yml` 
  - Run `yarn serverless-security --no-base-path` on another terminal window to start kibana in serverless mode
  - Run `node x-pack/plugins/security_solution/scripts/endpoint/endpoint_agent_emulator.js --asSuperuser` on a new window and then select `1` to `Load Endoints` and then `1` to `Run` the loader script. This will load some fake agents/alerts data to test with.

### Tests (Serverless)

- with 
`{ product_line: 'security', product_tier: 'essentials' }` or `{ product_line: 'security', product_tier: 'complete' }` 
and
`{ product_line: 'endpoint', product_tier: 'essentials' }` or `{ product_line: 'endpoint', product_tier: 'complete' }`

1.  Navigate to Rules>Shared exception lists via `http://localhost:5601/app/security/exceptions`
2. Test that you can see `Endpoint Security Exception List` card on the shared exception lists page.
3. Navigate to `Alerts` page via `app/security/alerts`, you should see endpoint alerts. If not, then click on `Manage Rules` and then disable/enable `Endpoint Security` rules. That should trigger alerts to show up on the Alerts table. 
4. Click on `View Details` button under `Actions` column. Once the flyout is visible, click on `Take Action` and verify that `Add Endpoint exception` is visible/enabled/clickable on the menu.
5. Click on `More actions` button under `Actions` column and verify that `Add Endpoint exception` is visible/enabled/clickable on the menu.
6. Click on `Investigate in timeline` button under `Actions` column; when the timeline view is visible and the alert item is displayed, click on buttons mentioned in 4. and 5. above and verify the same.
7. Navigate to `Rules`>`DetectionRules`>`Endpoint Security` rule under the `Rules` table.  Select the `Alerts` tab. 
8. Click and verify `View details`,`More actions` and `Investigate in timeline` buttons same as in 4., 5., 6. above.
9. You should be able to see the `Endpoint exceptions` tab as well. Click and verify that you can see the tab's content.

- with 
`{ product_line: 'security', product_tier: 'essentials' }` or `{ product_line: 'security', product_tier: 'complete' }`
1. Edit `config/serverless.security.dev.yml` so that `endpoint` product line item is commented out.
2. Test that you can not see `Endpoint Security Exception List` card on the shared exception lists page.
3. Items 4. 5. 6. as above but the menu items should be disabled. This can be verified with fake data only as with a real endpoint, endpoint alerts are actually not visible at all.


### Tests (ESS)
On the ESS side, endpoint exceptions are not affected by this change and work as usual based on index privileges.
